<a href="https://colab.research.google.com/github/MagicPolygon/spotify-data-analysis/blob/main/notebooks/data_preparation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Spotify Project

I'm doing a data analysis project on my Spotify Extended Streaming History Data. The analyses I hope to do are:

1. Total listening time over time
2. Top artists
3. Skip rate analysis

Initially, I planned to use Spotify's API to incorporate genre data into the whole dataset to do genre-related analysis, for example, Listening Time by Genre. But, I have time constraints, so this will be something to work on in the future.

I also wanted to analyse Time-of-day listening patterns, but it didn't quite fit the story of the analyses I already did.

This Jupyter Notebook attempts to get the data cleaned and ready to use for visualisation in Tableau.

## Imports and Mounting

To begin, I will import everything I need and mount Google Drive so that I can access the files there:

In [1]:
import os    # To enable the location of the project to be changed
import pandas as pd

from google.colab import drive    # To enable access to Google Drive files
from IPython.display import display    # To allow multiple outputs to be displayed in a single output cell

drive.mount("/content/drive")

Mounted at /content/drive


## Setting up the location to work in
I will now change the current working directory to be the project location:

In [2]:
project_location = "/content/drive/MyDrive/spotify-data-analysis"
os.chdir(project_location)

## Gathering audio locations
My Spotify Extended Streaming History data is in the form of multiple JSON files in a folder. There are 13 files, file 0 to file 12. I have already anonymised the data using my "anonymise" notebook. I will now put all the locations of the JSON files into one list:

In [3]:
audio_locations = [
    "data/anonymised/Streaming_History_Audio_2017-2020_0.json",
    "data/anonymised/Streaming_History_Audio_2020_1.json",
    "data/anonymised/Streaming_History_Audio_2020_2.json",
    "data/anonymised/Streaming_History_Audio_2020-2021_3.json",
    "data/anonymised/Streaming_History_Audio_2021_4.json",
    "data/anonymised/Streaming_History_Audio_2021_5.json",
    "data/anonymised/Streaming_History_Audio_2021-2022_6.json",
    "data/anonymised/Streaming_History_Audio_2022_7.json",
    "data/anonymised/Streaming_History_Audio_2022-2023_8.json",
    "data/anonymised/Streaming_History_Audio_2023_9.json",
    "data/anonymised/Streaming_History_Audio_2023-2024_10.json",
    "data/anonymised/Streaming_History_Audio_2024_11.json",
    "data/anonymised/Streaming_History_Audio_2024-2025_12.json"
]

## Creating a dataframe for each audio location with relevant columns
While I do have the locations of the JSON files, the data from them has not been read. So, I will read each of them into its own dataframe object. Additionally, there is information I'm not concerned with, so I will keep only the information (columns) that I care about.

Here is an explanation of each of the relevant columns:

1. ts:
    - Date and time of when the stream ended in UTC format.<br><br>
2. ms_played:
    - For how many milliseconds the track was played.<br><br>
3. master_metadata_track_name:
    - Name of the track.<br><br>
4. master_metadata_album_artist_name:
    - Name of the artist, band, or podcast.<br><br>
5. master_metadata_album_album_name:
    - Name of the album of the track.<br><br>
6. spotify_track_uri:
    - A Spotify Track URI, that is identifying the unique music track.<br><br>
7. episode_name:
    - Name of the episode of the podcast.<br><br>
8. episode_show_name:
    - Name of the show of the podcast.<br><br>
9. spotify_episode_uri:
    - A Spotify Episode URI, that is identifying the unique podcast episode.<br><br>
10. audiobook_title:
    - Title of the audiobook.<br><br>
11. audiobook_uri:
    - A Spotify Audiobook URI, that is identifying the unique audiobook.<br><br>
12. audiobook_chapter_uri:
    - A Spotify Audiobook Chapter URI, that is identifying the unique audiobook chapter.<br><br>
13. audiobook_chapter_title:
    - Title of the audiobook chapter.<br><br>
14. reason_start:
    - Reason why the track started (e.g. previous track finished or you picked it from the playlist).<br><br>
15. reason_end:
    - Reason why the track ended (e.g. the track finished playing or you hit the next button).<br><br>
16. shuffle:
    - Whether shuffle mode was used when playing the track.<br><br>
17. skipped:
    - Information whether the user skipped to the next song.<br><br>

I am analysing tracks, not podcasts or audiobooks. But, I will inlcude columns for podcasts episodes and audiobooks in case there will be any podcast episodes or audiobooks in the dataframes. They will be removed. This will also account for the case of tracks that are also podcast episodes or audiobooks, if that's possible. If there are any, they will also be removed.

Finally, I will put all of the dataframes together into one list:

In [4]:
audio_dfs = []
relevant_columns = [
    "ts",
    "ms_played",
    "master_metadata_track_name",
    "master_metadata_album_artist_name",
    "master_metadata_album_album_name",
    "spotify_track_uri",
    "episode_name",
    "episode_show_name",
    "spotify_episode_uri",
    "audiobook_title",
    "audiobook_uri",
    "audiobook_chapter_uri",
    "audiobook_chapter_title",
    "reason_start",
    "reason_end",
    "shuffle",
    "skipped"
]

for audio_location in audio_locations:    # For each JSON file, read it, select only the relevant columns, then add it to a list
    audio_df = pd.read_json(audio_location)
    audio_df = audio_df[relevant_columns].copy()
    audio_dfs.append(audio_df)

## Learning general info about each dataframe
I need to have an idea of what each dataframe is like so that I can eventually put them together nicely and have them ready for analysis, but I'm unsure of the best way to do this. I need to start somewhere, so for each dataframe, I will look at some general information:

In [5]:
for audio_df in audio_dfs:
    audio_df.info()
    print("\n")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15579 entries, 0 to 15578
Data columns (total 17 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   ts                                 15579 non-null  object 
 1   ms_played                          15579 non-null  int64  
 2   master_metadata_track_name         15579 non-null  object 
 3   master_metadata_album_artist_name  15579 non-null  object 
 4   master_metadata_album_album_name   15579 non-null  object 
 5   spotify_track_uri                  15579 non-null  object 
 6   episode_name                       0 non-null      float64
 7   episode_show_name                  0 non-null      float64
 8   spotify_episode_uri                0 non-null      float64
 9   audiobook_title                    0 non-null      float64
 10  audiobook_uri                      0 non-null      float64
 11  audiobook_chapter_uri              0 non-null      flo

## Checking that the column names of each dataframe are equal
Comparing the above general information for each dataframe against each other is not easy, so I'll also directly compare the columns of each dataframe:

In [6]:
print("The column names are the same for dataframe")
for index, audio_df in enumerate(audio_dfs):
    if index != len(audio_dfs) - 1:    # Ensuring I don't go out of bounds for the list
        if audio_df.columns.equals(audio_dfs[index + 1].columns):
            print(f"{index} and {index + 1}? Yes.")
        else:
            print(f"{index} and {index + 1}? No.")

The column names are the same for dataframe
0 and 1? Yes.
1 and 2? Yes.
2 and 3? Yes.
3 and 4? Yes.
4 and 5? Yes.
5 and 6? Yes.
6 and 7? Yes.
7 and 8? Yes.
8 and 9? Yes.
9 and 10? Yes.
10 and 11? Yes.
11 and 12? Yes.


## Checking that the data types are consistent
Now, I will do that same as above, except for the data type of each column instead of for column names:

In [7]:
print("The data types are consistent for dataframe")
for index, audio_df in enumerate(audio_dfs):
    if index != len(audio_dfs) - 1:    # Ensuring I don't go out of bounds for the list
        if audio_df.dtypes.equals(audio_dfs[index + 1].dtypes):
            print(f"{index} and {index + 1}? Yes.")
        else:
            print(f"{index} and {index + 1}? No.")

The data types are consistent for dataframe
0 and 1? Yes.
1 and 2? Yes.
2 and 3? Yes.
3 and 4? Yes.
4 and 5? Yes.
5 and 6? Yes.
6 and 7? Yes.
7 and 8? Yes.
8 and 9? Yes.
9 and 10? Yes.
10 and 11? Yes.
11 and 12? No.


## Investigating the difference in data types between dataframe 11 and 12

From above, I noticed that the data types are consistent everywhere, except for dataframe 11 and 12. Upon investigating the general information for dataframe 11 and 12 further above, I learned the differences in data type. It regards the episode_name, episode_show_name, and spotify_episode_uri columns. For dataframes 0 to 11, all values in each of these columns are null, and the data type is float64. But for dataframe 12, in each of these columns, there is 1 non-null value. Also, the data type of those columns is object instead of float64.

This may suggest there is a podcast episode in this dataframe, but I can't be sure. My analyses are for tracks, not podcast episodes, so if there are any episodes, they need to be removed.

## Checking distinct values for podcast episode columns and audiobook columns
So, there is podcast episode information in dataframe 12. I want to be absolutely certain there is no podcast or audiobook information in any of my dataframes, so I will go through each dataframe, and each of their podcast episode and audiobook columns to check for their unique values:

In [8]:
questionable_columns = [
    "episode_name",
    "episode_show_name",
    "spotify_episode_uri",
    "audiobook_title",
    "audiobook_uri",
    "audiobook_chapter_uri",
    "audiobook_chapter_title"
]

for index, audio_df in enumerate(audio_dfs):
    for questionable_column in questionable_columns:
        print(f"DataFrame: \t{index}\nColumn: \t{questionable_column}")
        display(audio_df[questionable_column].value_counts())
        print("\n")

DataFrame: 	0
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	0
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	0
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	0
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	0
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	0
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	0
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	1
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	1
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	1
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	1
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	1
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	1
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	1
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	2
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	2
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	2
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	2
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	2
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	2
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	2
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	3
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	3
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	3
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	3
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	3
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	3
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	3
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	4
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	4
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	4
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	4
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	4
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	4
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	4
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	5
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	5
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	5
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	5
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	5
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	5
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	5
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	6
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	6
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	6
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	6
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	6
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	6
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	6
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	7
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	7
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	7
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	7
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	7
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	7
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	7
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	8
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	8
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	8
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	8
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	8
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	8
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	8
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	9
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	9
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	9
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	9
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	9
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	9
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	9
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	10
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	10
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	10
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	10
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	10
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	10
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	10
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	11
Column: 	episode_name


Series([], Name: count, dtype: int64)



DataFrame: 	11
Column: 	episode_show_name


Series([], Name: count, dtype: int64)



DataFrame: 	11
Column: 	spotify_episode_uri


Series([], Name: count, dtype: int64)



DataFrame: 	11
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	11
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	11
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	11
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)



DataFrame: 	12
Column: 	episode_name


episode_name
76. The ULEZ Blade Runners (the fuming suburbanites chopping down ULEZ cameras)    1
Name: count, dtype: int64



DataFrame: 	12
Column: 	episode_show_name


episode_show_name
The Screen Rot Podcast with Jacob and Jake    1
Name: count, dtype: int64



DataFrame: 	12
Column: 	spotify_episode_uri


spotify_episode_uri
spotify:episode:7AZTPXZTmoeGJKwXCgl6g6    1
Name: count, dtype: int64



DataFrame: 	12
Column: 	audiobook_title


Series([], Name: count, dtype: int64)



DataFrame: 	12
Column: 	audiobook_uri


Series([], Name: count, dtype: int64)



DataFrame: 	12
Column: 	audiobook_chapter_uri


Series([], Name: count, dtype: int64)



DataFrame: 	12
Column: 	audiobook_chapter_title


Series([], Name: count, dtype: int64)

## Learning about the episode information in dataframe 12

From the above, we see that for each dataframe, and for each of their podcast episode columns and audiobook columns, there are no values, except for dataframe 12. Here, there is one value in its episode_name column, one in episode_show_name, and one in spotify_episode_uri. This is decribed below.

Audio DataFrame: 12, Column: episode_name
episode_name
76. The ULEZ Blade Runners (the fuming suburbanites chopping down ULEZ cameras)    1
Name: count, dtype: int64

Audio DataFrame: 12, Column: episode_show_name
episode_show_name
The Screen Rot Podcast with Jacob and Jake    1
Name: count, dtype: int64

Audio DataFrame: 12, Column: spotify_episode_uri
spotify_episode_uri
spotify:episode:7AZTPXZTmoeGJKwXCgl6g6    1
Name: count, dtype: int64

## Finding the episode in dataframe 12

Now, I need to find where this information lives. I imagine that the above values all live in the same row, so I will look at the row in dataframe 12 that contains the value in the episode_name column to see if the episode_show_name value and the spotify_episode_uri are also in that row:

In [9]:
audio_df_12 = audio_dfs[12]
episode_name_column = audio_df_12["episode_name"].copy()
audio_df_12[episode_name_column.notna()]

ts  ms_played master_metadata_track_name  \
5090  2025-03-09T20:37:49Z          0                       None   

     master_metadata_album_artist_name master_metadata_album_album_name  \
5090                              None                             None   

     spotify_track_uri                                       episode_name  \
5090              None  76. The ULEZ Blade Runners (the fuming suburba...   

                               episode_show_name  \
5090  The Screen Rot Podcast with Jacob and Jake   

                         spotify_episode_uri  audiobook_title  audiobook_uri  \
5090  spotify:episode:7AZTPXZTmoeGJKwXCgl6g6              NaN            NaN   

      audiobook_chapter_uri  audiobook_chapter_title reason_start reason_end  \
5090                    NaN                      NaN      unknown    endplay   

      shuffle  skipped  
5090    False     True

## Removing the episode

As we can see, the row that contains the episode_name value also contains the episode_show_name value and the spotify_episode_uri value. This row doesn't have track data, so it's not a track. I will safely remove this:

In [10]:
audio_track_df_12 = audio_df_12[episode_name_column.isna()].copy()
audio_track_df_12

ts  ms_played master_metadata_track_name  \
0      2024-12-20T23:16:07Z     178051                       Iron   
1      2024-12-20T23:20:11Z     243304                Galaxy Wars   
2      2024-12-20T23:27:21Z     163377                  Like this   
3      2024-12-21T14:27:38Z      74626                    Entreat   
4      2024-12-21T14:28:30Z      39973                  Like this   
...                     ...        ...                        ...   
12101  2025-08-16T19:31:58Z      36896                Fukurokouji   
12102  2025-08-16T19:32:03Z       4563                Playstation   
12103  2025-08-16T19:38:07Z     364000                Fukurokouji   
12104  2025-08-16T19:41:14Z     186227               Door Shutter   
12105  2025-08-16T19:46:19Z     304080                     Locked   

      master_metadata_album_artist_name master_metadata_album_album_name  \
0                               Woodkid                             Iron   
1                           Evil Needle                           Qualia   
2                      박혜진 Park Hye Jin                        How can I   
3                  ||||||||||||||||||||                          Entreat   
4                      박혜진 Park Hye Jin                        How can I   
...                                 ...                              ...   
12101                         El Huervo                      World's End   
12102                          Ebi Soda                              Ugh   
12103                         El Huervo                      World's End   
12104                       Brokenchord             Endless Transmission   
12105                           Dimlite                This Is Embracing   

                          spotify_track_uri episode_name episode_show_name  \
0      spotify:track:0j1VVFif10Ib9s3U59adli         None              None   
1      spotify:track:1FRnIWyCmEk38q1TsZhlCA         None              None   
2      spotify:track:7pY3HE16Ir4ZFTEOmosP57         None              None   
3      spotify:track:7BrBhLOpXQma7QpvnWxAvZ         None              None   
4      spotify:track:7pY3HE16Ir4ZFTEOmosP57         None              None   
...                                     ...          ...               ...   
12101  spotify:track:771NtAeF8EiFO4D4CQG2ij         None              None   
12102  spotify:track:5PETadLT856KzshN8ArpWl         None              None   
12103  spotify:track:771NtAeF8EiFO4D4CQG2ij         None              None   
12104  spotify:track:2bBrirDuZf2mDbE6P7PCwh         None              None   
12105  spotify:track:3To51zCuaPkUK9HoUOXHfI         None              None   

      spotify_episode_uri  audiobook_title  audiobook_uri  \
0                    None              NaN            NaN   
1                    None              NaN            NaN   
2                    None              NaN            NaN   
3                    None              NaN            NaN   
4                    None              NaN            NaN   
...                   ...              ...            ...   
12101                None              NaN            NaN   
12102                None              NaN            NaN   
12103                None              NaN            NaN   
12104                None              NaN            NaN   
12105                None              NaN            NaN   

       audiobook_chapter_uri  audiobook_chapter_title reason_start reason_end  \
0                        NaN                      NaN    trackdone  trackdone   
1                        NaN                      NaN    trackdone  trackdone   
2                        NaN                      NaN    trackdone     logout   
3                        NaN                      NaN    trackdone     remote   
4                        NaN                      NaN       remote    endplay   
...                      ...                      ...          ...        ...   
12101                    NaN                  

## Missing index

I noticed that in the above dataframe, there are 12105 row, but the first index is 0, and the last index is 12105. This means that the indices of the rows that come after the removed episode row aren't updated. We will do this:

In [11]:
audio_track_df_12 = audio_track_df_12.reset_index(drop=True)    # drop=True, so the old index doesn't become a column
audio_track_df_12

ts  ms_played master_metadata_track_name  \
0      2024-12-20T23:16:07Z     178051                       Iron   
1      2024-12-20T23:20:11Z     243304                Galaxy Wars   
2      2024-12-20T23:27:21Z     163377                  Like this   
3      2024-12-21T14:27:38Z      74626                    Entreat   
4      2024-12-21T14:28:30Z      39973                  Like this   
...                     ...        ...                        ...   
12100  2025-08-16T19:31:58Z      36896                Fukurokouji   
12101  2025-08-16T19:32:03Z       4563                Playstation   
12102  2025-08-16T19:38:07Z     364000                Fukurokouji   
12103  2025-08-16T19:41:14Z     186227               Door Shutter   
12104  2025-08-16T19:46:19Z     304080                     Locked   

      master_metadata_album_artist_name master_metadata_album_album_name  \
0                               Woodkid                             Iron   
1                           Evil Needle                           Qualia   
2                      박혜진 Park Hye Jin                        How can I   
3                  ||||||||||||||||||||                          Entreat   
4                      박혜진 Park Hye Jin                        How can I   
...                                 ...                              ...   
12100                         El Huervo                      World's End   
12101                          Ebi Soda                              Ugh   
12102                         El Huervo                      World's End   
12103                       Brokenchord             Endless Transmission   
12104                           Dimlite                This Is Embracing   

                          spotify_track_uri episode_name episode_show_name  \
0      spotify:track:0j1VVFif10Ib9s3U59adli         None              None   
1      spotify:track:1FRnIWyCmEk38q1TsZhlCA         None              None   
2      spotify:track:7pY3HE16Ir4ZFTEOmosP57         None              None   
3      spotify:track:7BrBhLOpXQma7QpvnWxAvZ         None              None   
4      spotify:track:7pY3HE16Ir4ZFTEOmosP57         None              None   
...                                     ...          ...               ...   
12100  spotify:track:771NtAeF8EiFO4D4CQG2ij         None              None   
12101  spotify:track:5PETadLT856KzshN8ArpWl         None              None   
12102  spotify:track:771NtAeF8EiFO4D4CQG2ij         None              None   
12103  spotify:track:2bBrirDuZf2mDbE6P7PCwh         None              None   
12104  spotify:track:3To51zCuaPkUK9HoUOXHfI         None              None   

      spotify_episode_uri  audiobook_title  audiobook_uri  \
0                    None              NaN            NaN   
1                    None              NaN            NaN   
2                    None              NaN            NaN   
3                    None              NaN            NaN   
4                    None              NaN            NaN   
...                   ...              ...            ...   
12100                None              NaN            NaN   
12101                None              NaN            NaN   
12102                None              NaN            NaN   
12103                None              NaN            NaN   
12104                None              NaN            NaN   

       audiobook_chapter_uri  audiobook_chapter_title reason_start reason_end  \
0                        NaN                      NaN    trackdone  trackdone   
1                        NaN                      NaN    trackdone  trackdone   
2                        NaN                      NaN    trackdone     logout   
3                        NaN                      NaN    trackdone     remote   
4                        NaN                      NaN       remote    endplay   
...                      ...                      ...          ...        ...   
12100                    NaN                  

## Assigning the neatened dataframe 12 back to the list of dataframes

Now, the final index is 12104, rather than 12105, meaning the indices have been updated to account for the gap. We now have to update the dataframe list so that its 12th entry is replaced with its version that's purely comprised of tracks.

In [12]:
audio_dfs[12] = audio_track_df_12
audio_dfs[12]

ts  ms_played master_metadata_track_name  \
0      2024-12-20T23:16:07Z     178051                       Iron   
1      2024-12-20T23:20:11Z     243304                Galaxy Wars   
2      2024-12-20T23:27:21Z     163377                  Like this   
3      2024-12-21T14:27:38Z      74626                    Entreat   
4      2024-12-21T14:28:30Z      39973                  Like this   
...                     ...        ...                        ...   
12100  2025-08-16T19:31:58Z      36896                Fukurokouji   
12101  2025-08-16T19:32:03Z       4563                Playstation   
12102  2025-08-16T19:38:07Z     364000                Fukurokouji   
12103  2025-08-16T19:41:14Z     186227               Door Shutter   
12104  2025-08-16T19:46:19Z     304080                     Locked   

      master_metadata_album_artist_name master_metadata_album_album_name  \
0                               Woodkid                             Iron   
1                           Evil Needle                           Qualia   
2                      박혜진 Park Hye Jin                        How can I   
3                  ||||||||||||||||||||                          Entreat   
4                      박혜진 Park Hye Jin                        How can I   
...                                 ...                              ...   
12100                         El Huervo                      World's End   
12101                          Ebi Soda                              Ugh   
12102                         El Huervo                      World's End   
12103                       Brokenchord             Endless Transmission   
12104                           Dimlite                This Is Embracing   

                          spotify_track_uri episode_name episode_show_name  \
0      spotify:track:0j1VVFif10Ib9s3U59adli         None              None   
1      spotify:track:1FRnIWyCmEk38q1TsZhlCA         None              None   
2      spotify:track:7pY3HE16Ir4ZFTEOmosP57         None              None   
3      spotify:track:7BrBhLOpXQma7QpvnWxAvZ         None              None   
4      spotify:track:7pY3HE16Ir4ZFTEOmosP57         None              None   
...                                     ...          ...               ...   
12100  spotify:track:771NtAeF8EiFO4D4CQG2ij         None              None   
12101  spotify:track:5PETadLT856KzshN8ArpWl         None              None   
12102  spotify:track:771NtAeF8EiFO4D4CQG2ij         None              None   
12103  spotify:track:2bBrirDuZf2mDbE6P7PCwh         None              None   
12104  spotify:track:3To51zCuaPkUK9HoUOXHfI         None              None   

      spotify_episode_uri  audiobook_title  audiobook_uri  \
0                    None              NaN            NaN   
1                    None              NaN            NaN   
2                    None              NaN            NaN   
3                    None              NaN            NaN   
4                    None              NaN            NaN   
...                   ...              ...            ...   
12100                None              NaN            NaN   
12101                None              NaN            NaN   
12102                None              NaN            NaN   
12103                None              NaN            NaN   
12104                None              NaN            NaN   

       audiobook_chapter_uri  audiobook_chapter_title reason_start reason_end  \
0                        NaN                      NaN    trackdone  trackdone   
1                        NaN                      NaN    trackdone  trackdone   
2                        NaN                      NaN    trackdone     logout   
3                        NaN                      NaN    trackdone     remote   
4                        NaN                      NaN       remote    endplay   
...                      ...                      ...          ...        ...   
12100                    NaN                  

## Keeping the relevant columns of each dataframe

All the dataframes purely made up of tracks now, and all podcast episode columns and audiobook columns have no values. I feel safe changing my dataframes so they're only made of the columns that may be needed for my analyses:

In [13]:
final_relevant_columns = [
    "ts",
    "ms_played",
    "master_metadata_track_name",
    "master_metadata_album_artist_name",
    "master_metadata_album_album_name",
    "spotify_track_uri",
    "reason_start",
    "reason_end",
    "shuffle",
    "skipped"
]

for index, audio_df in enumerate(audio_dfs):
    audio_dfs[index] = audio_df[final_relevant_columns].copy()

## Checking the general information of each dataframe again

Because of all of the changes I've made already, I want to be sure that each dataframe looks OK, so I'll look at their general information again.

In [14]:
for audio_df in audio_dfs:
    audio_df.info()
    print("\n")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15579 entries, 0 to 15578
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   ts                                 15579 non-null  object
 1   ms_played                          15579 non-null  int64 
 2   master_metadata_track_name         15579 non-null  object
 3   master_metadata_album_artist_name  15579 non-null  object
 4   master_metadata_album_album_name   15579 non-null  object
 5   spotify_track_uri                  15579 non-null  object
 6   reason_start                       15579 non-null  object
 7   reason_end                         15579 non-null  object
 8   shuffle                            15579 non-null  bool  
 9   skipped                            15579 non-null  bool  
dtypes: bool(2), int64(1), object(7)
memory usage: 1004.2+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15469 entries, 0 to 1

## Investigating the type of ts values

I noticed that object is the Dtype of all ts columns. I expected them to come out as dates, since ts, described by Spotify, is a date and time. So, I will learn about the type of all ts values to have a better idea of what is going on.

In [15]:
for audio_df in audio_dfs:
    ts_values = audio_df["ts"].copy()
    ts_types = ts_values.apply(type)    # Use the type function to produce a Series of types for each ts value
    ts_types_count = ts_types.value_counts()    # Count how many of each type appears
    display(ts_types_count)

ts
<class 'str'>    15579
Name: count, dtype: int64

ts
<class 'str'>    15469
Name: count, dtype: int64

ts
<class 'str'>    15540
Name: count, dtype: int64

ts
<class 'str'>    15514
Name: count, dtype: int64

ts
<class 'str'>    15340
Name: count, dtype: int64

ts
<class 'str'>    15266
Name: count, dtype: int64

ts
<class 'str'>    15362
Name: count, dtype: int64

ts
<class 'str'>    15552
Name: count, dtype: int64

ts
<class 'str'>    16124
Name: count, dtype: int64

ts
<class 'str'>    16149
Name: count, dtype: int64

ts
<class 'str'>    16123
Name: count, dtype: int64

ts
<class 'str'>    16102
Name: count, dtype: int64

ts
<class 'str'>    12105
Name: count, dtype: int64

## Testing the ts strings to see if they're dates

We see that all of the ts values are strings. After looking at many displays of the data, the ts values seem to take the format of the following example: `2017-04-01T17:43:10Z`. This is an ISO 8601 datetime string. This can be parsed from a string to a datetime. While, I don't know what all ts values look like, I can attempt to parse all values into a datetime and be notified if any ts value is not a parseable timestamp using `pd.to_datetime(..., errors='raise')`:

In [16]:
for audio_df in audio_dfs:
    audio_df["ts"] = pd.to_datetime(audio_df["ts"], errors="raise", utc=True)    # utc=True ensures the datetime is timezone-aware and in UTC

## Investigating the type of ts values again

No error were raised. I will check the types of each ts value again like before.


In [17]:
for audio_df in audio_dfs:
    ts_values = audio_df["ts"].copy()
    ts_types = ts_values.apply(type)    # Use the type function to produce a Series of types for each ts value
    ts_types_count = ts_types.value_counts()    # Count how many of each type appears
    display(ts_types_count)

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    15579
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    15469
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    15540
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    15514
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    15340
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    15266
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    15362
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    15552
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    16124
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    16149
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    16123
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    16102
Name: count, dtype: int64

ts
<class 'pandas._libs.tslibs.timestamps.Timestamp'>    12105
Name: count, dtype: int64

## Checking the data type of the ts columns themselves

So, we see that the type of values in the ts columns are all `pandas._libs.tslibs.timestamps.Timestamp`. This corresponds to `datetime64[ns, UTC]`, which is what we're hoping for. Now, I will check the columns that actually contain these values.

In [18]:
for audio_df in audio_dfs:
  display(audio_df["ts"].dtype)

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

datetime64[ns, UTC]

## Checking ts nulls

Now that I'm happy with the ts data type, I want to look at info about all the dataframes again.

In [19]:
for audio_df in audio_dfs:
    audio_df.info()
    print("\n")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15579 entries, 0 to 15578
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15579 non-null  datetime64[ns, UTC]
 1   ms_played                          15579 non-null  int64              
 2   master_metadata_track_name         15579 non-null  object             
 3   master_metadata_album_artist_name  15579 non-null  object             
 4   master_metadata_album_album_name   15579 non-null  object             
 5   spotify_track_uri                  15579 non-null  object             
 6   reason_start                       15579 non-null  object             
 7   reason_end                         15579 non-null  object             
 8   shuffle                            15579 non-null  bool               
 9   skipped                            15579 non-null 

## Object Dtype concerns

We can see that for the columns

- master_metadata_track_name,
- master_metadata_album_artist_name,
- master_metadata_album_album_name,
- spotify_track_uri,
- reason_start,
-  and reason_end,

their Dtype are all objects. I expected them to be strings. So, before converting the column Dtypes to string, I will first inspect if all the values in all these columns are strings.

In [20]:
object_columns = [
    "master_metadata_track_name",
    "master_metadata_album_artist_name",
    "master_metadata_album_album_name",
    "spotify_track_uri",
    "reason_start",
    "reason_end"
]

for audio_df in audio_dfs:
  print(len(audio_df))
  for object_column in object_columns:
    display(audio_df[object_column].apply(type).value_counts())

15579


master_metadata_track_name
<class 'str'>    15579
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    15579
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    15579
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    15579
Name: count, dtype: int64

reason_start
<class 'str'>    15579
Name: count, dtype: int64

reason_end
<class 'str'>    15579
Name: count, dtype: int64

15469


master_metadata_track_name
<class 'str'>    15469
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    15469
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    15469
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    15469
Name: count, dtype: int64

reason_start
<class 'str'>    15469
Name: count, dtype: int64

reason_end
<class 'str'>    15469
Name: count, dtype: int64

15540


master_metadata_track_name
<class 'str'>    15540
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    15540
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    15540
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    15540
Name: count, dtype: int64

reason_start
<class 'str'>    15540
Name: count, dtype: int64

reason_end
<class 'str'>    15540
Name: count, dtype: int64

15514


master_metadata_track_name
<class 'str'>    15514
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    15514
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    15514
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    15514
Name: count, dtype: int64

reason_start
<class 'str'>    15514
Name: count, dtype: int64

reason_end
<class 'str'>    15514
Name: count, dtype: int64

15340


master_metadata_track_name
<class 'str'>    15340
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    15340
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    15340
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    15340
Name: count, dtype: int64

reason_start
<class 'str'>    15340
Name: count, dtype: int64

reason_end
<class 'str'>    15340
Name: count, dtype: int64

15266


master_metadata_track_name
<class 'str'>    15266
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    15266
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    15266
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    15266
Name: count, dtype: int64

reason_start
<class 'str'>    15266
Name: count, dtype: int64

reason_end
<class 'str'>    15266
Name: count, dtype: int64

15362


master_metadata_track_name
<class 'str'>    15362
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    15362
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    15362
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    15362
Name: count, dtype: int64

reason_start
<class 'str'>    15362
Name: count, dtype: int64

reason_end
<class 'str'>    15362
Name: count, dtype: int64

15552


master_metadata_track_name
<class 'str'>    15552
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    15552
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    15552
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    15552
Name: count, dtype: int64

reason_start
<class 'str'>    15552
Name: count, dtype: int64

reason_end
<class 'str'>    15552
Name: count, dtype: int64

16124


master_metadata_track_name
<class 'str'>    16124
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    16124
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    16124
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    16124
Name: count, dtype: int64

reason_start
<class 'str'>    16124
Name: count, dtype: int64

reason_end
<class 'str'>    16124
Name: count, dtype: int64

16149


master_metadata_track_name
<class 'str'>    16149
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    16149
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    16149
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    16149
Name: count, dtype: int64

reason_start
<class 'str'>    16149
Name: count, dtype: int64

reason_end
<class 'str'>    16149
Name: count, dtype: int64

16123


master_metadata_track_name
<class 'str'>    16123
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    16123
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    16123
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    16123
Name: count, dtype: int64

reason_start
<class 'str'>    16123
Name: count, dtype: int64

reason_end
<class 'str'>    16123
Name: count, dtype: int64

16102


master_metadata_track_name
<class 'str'>    16102
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    16102
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    16102
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    16102
Name: count, dtype: int64

reason_start
<class 'str'>    16102
Name: count, dtype: int64

reason_end
<class 'str'>    16102
Name: count, dtype: int64

12105


master_metadata_track_name
<class 'str'>    12105
Name: count, dtype: int64

master_metadata_album_artist_name
<class 'str'>    12105
Name: count, dtype: int64

master_metadata_album_album_name
<class 'str'>    12105
Name: count, dtype: int64

spotify_track_uri
<class 'str'>    12105
Name: count, dtype: int64

reason_start
<class 'str'>    12105
Name: count, dtype: int64

reason_end
<class 'str'>    12105
Name: count, dtype: int64

## Converting the columns to string columns

From the above, we see that they are all strings. I will now convert those columns to string columns.

In [21]:
for audio_df in audio_dfs:
  for object_column in object_columns:
    audio_df[object_column] = audio_df[object_column].astype("string")

## Checking the dataframes
I want to make sure that they colmumns are now string columns.

In [22]:
for audio_df in audio_dfs:
  display(audio_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15579 entries, 0 to 15578
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15579 non-null  datetime64[ns, UTC]
 1   ms_played                          15579 non-null  int64              
 2   master_metadata_track_name         15579 non-null  string             
 3   master_metadata_album_artist_name  15579 non-null  string             
 4   master_metadata_album_album_name   15579 non-null  string             
 5   spotify_track_uri                  15579 non-null  string             
 6   reason_start                       15579 non-null  string             
 7   reason_end                         15579 non-null  string             
 8   shuffle                            15579 non-null  bool               
 9   skipped                            15579 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15469 entries, 0 to 15468
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15469 non-null  datetime64[ns, UTC]
 1   ms_played                          15469 non-null  int64              
 2   master_metadata_track_name         15469 non-null  string             
 3   master_metadata_album_artist_name  15469 non-null  string             
 4   master_metadata_album_album_name   15469 non-null  string             
 5   spotify_track_uri                  15469 non-null  string             
 6   reason_start                       15469 non-null  string             
 7   reason_end                         15469 non-null  string             
 8   shuffle                            15469 non-null  bool               
 9   skipped                            15469 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15540 entries, 0 to 15539
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15540 non-null  datetime64[ns, UTC]
 1   ms_played                          15540 non-null  int64              
 2   master_metadata_track_name         15540 non-null  string             
 3   master_metadata_album_artist_name  15540 non-null  string             
 4   master_metadata_album_album_name   15540 non-null  string             
 5   spotify_track_uri                  15540 non-null  string             
 6   reason_start                       15540 non-null  string             
 7   reason_end                         15540 non-null  string             
 8   shuffle                            15540 non-null  bool               
 9   skipped                            15540 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15514 entries, 0 to 15513
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15514 non-null  datetime64[ns, UTC]
 1   ms_played                          15514 non-null  int64              
 2   master_metadata_track_name         15514 non-null  string             
 3   master_metadata_album_artist_name  15514 non-null  string             
 4   master_metadata_album_album_name   15514 non-null  string             
 5   spotify_track_uri                  15514 non-null  string             
 6   reason_start                       15514 non-null  string             
 7   reason_end                         15514 non-null  string             
 8   shuffle                            15514 non-null  bool               
 9   skipped                            15514 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15340 entries, 0 to 15339
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15340 non-null  datetime64[ns, UTC]
 1   ms_played                          15340 non-null  int64              
 2   master_metadata_track_name         15340 non-null  string             
 3   master_metadata_album_artist_name  15340 non-null  string             
 4   master_metadata_album_album_name   15340 non-null  string             
 5   spotify_track_uri                  15340 non-null  string             
 6   reason_start                       15340 non-null  string             
 7   reason_end                         15340 non-null  string             
 8   shuffle                            15340 non-null  bool               
 9   skipped                            15340 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15266 entries, 0 to 15265
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15266 non-null  datetime64[ns, UTC]
 1   ms_played                          15266 non-null  int64              
 2   master_metadata_track_name         15266 non-null  string             
 3   master_metadata_album_artist_name  15266 non-null  string             
 4   master_metadata_album_album_name   15266 non-null  string             
 5   spotify_track_uri                  15266 non-null  string             
 6   reason_start                       15266 non-null  string             
 7   reason_end                         15266 non-null  string             
 8   shuffle                            15266 non-null  bool               
 9   skipped                            15266 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15362 entries, 0 to 15361
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15362 non-null  datetime64[ns, UTC]
 1   ms_played                          15362 non-null  int64              
 2   master_metadata_track_name         15362 non-null  string             
 3   master_metadata_album_artist_name  15362 non-null  string             
 4   master_metadata_album_album_name   15362 non-null  string             
 5   spotify_track_uri                  15362 non-null  string             
 6   reason_start                       15362 non-null  string             
 7   reason_end                         15362 non-null  string             
 8   shuffle                            15362 non-null  bool               
 9   skipped                            15362 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15552 entries, 0 to 15551
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 15552 non-null  datetime64[ns, UTC]
 1   ms_played                          15552 non-null  int64              
 2   master_metadata_track_name         15552 non-null  string             
 3   master_metadata_album_artist_name  15552 non-null  string             
 4   master_metadata_album_album_name   15552 non-null  string             
 5   spotify_track_uri                  15552 non-null  string             
 6   reason_start                       15552 non-null  string             
 7   reason_end                         15552 non-null  string             
 8   shuffle                            15552 non-null  bool               
 9   skipped                            15552 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16124 entries, 0 to 16123
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 16124 non-null  datetime64[ns, UTC]
 1   ms_played                          16124 non-null  int64              
 2   master_metadata_track_name         16124 non-null  string             
 3   master_metadata_album_artist_name  16124 non-null  string             
 4   master_metadata_album_album_name   16124 non-null  string             
 5   spotify_track_uri                  16124 non-null  string             
 6   reason_start                       16124 non-null  string             
 7   reason_end                         16124 non-null  string             
 8   shuffle                            16124 non-null  bool               
 9   skipped                            16124 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16149 entries, 0 to 16148
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 16149 non-null  datetime64[ns, UTC]
 1   ms_played                          16149 non-null  int64              
 2   master_metadata_track_name         16149 non-null  string             
 3   master_metadata_album_artist_name  16149 non-null  string             
 4   master_metadata_album_album_name   16149 non-null  string             
 5   spotify_track_uri                  16149 non-null  string             
 6   reason_start                       16149 non-null  string             
 7   reason_end                         16149 non-null  string             
 8   shuffle                            16149 non-null  bool               
 9   skipped                            16149 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16123 entries, 0 to 16122
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 16123 non-null  datetime64[ns, UTC]
 1   ms_played                          16123 non-null  int64              
 2   master_metadata_track_name         16123 non-null  string             
 3   master_metadata_album_artist_name  16123 non-null  string             
 4   master_metadata_album_album_name   16123 non-null  string             
 5   spotify_track_uri                  16123 non-null  string             
 6   reason_start                       16123 non-null  string             
 7   reason_end                         16123 non-null  string             
 8   shuffle                            16123 non-null  bool               
 9   skipped                            16123 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16102 entries, 0 to 16101
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 16102 non-null  datetime64[ns, UTC]
 1   ms_played                          16102 non-null  int64              
 2   master_metadata_track_name         16102 non-null  string             
 3   master_metadata_album_artist_name  16102 non-null  string             
 4   master_metadata_album_album_name   16102 non-null  string             
 5   spotify_track_uri                  16102 non-null  string             
 6   reason_start                       16102 non-null  string             
 7   reason_end                         16102 non-null  string             
 8   shuffle                            16102 non-null  bool               
 9   skipped                            16102 non-null 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12105 entries, 0 to 12104
Data columns (total 10 columns):
 #   Column                             Non-Null Count  Dtype              
---  ------                             --------------  -----              
 0   ts                                 12105 non-null  datetime64[ns, UTC]
 1   ms_played                          12105 non-null  int64              
 2   master_metadata_track_name         12105 non-null  string             
 3   master_metadata_album_artist_name  12105 non-null  string             
 4   master_metadata_album_album_name   12105 non-null  string             
 5   spotify_track_uri                  12105 non-null  string             
 6   reason_start                       12105 non-null  string             
 7   reason_end                         12105 non-null  string             
 8   shuffle                            12105 non-null  bool               
 9   skipped                            12105 non-null 

None

## Bringing the dataframes together into a single dataframe

I feel like I've done enough for the independent dataframes to allow me to bring them together without much issue. I will do this:

In [23]:
actual_num_of_rows = 0

for audio_df in audio_dfs:
  actual_num_of_rows += len(audio_df)

print(f"The total number of rows in the combined dataframe should be {actual_num_of_rows}")

audio_df = pd.concat(audio_dfs, ignore_index=True)    # ignore_index automatically reindexes the concatenated dataframe
calulcated_num_of_rows = len(audio_df)
print(f"The calulcated number of rows is {calulcated_num_of_rows}")

audio_df

The total number of rows in the combined dataframe should be 200225
The calulcated number of rows is 200225


ts  ms_played   master_metadata_track_name  \
0      2017-04-01 17:43:10+00:00      36860  Romance for Cello and Piano   
1      2017-04-01 19:52:27+00:00      36800               Pavane, Op. 50   
2      2017-04-01 19:57:07+00:00     280933          Funkin' for Jamaica   
3      2017-04-01 20:01:58+00:00     290160           Love Rollercoaster   
4      2017-04-01 20:09:38+00:00     459533                  The Payback   
...                          ...        ...                          ...   
200220 2025-08-16 19:31:58+00:00      36896                  Fukurokouji   
200221 2025-08-16 19:32:03+00:00       4563                  Playstation   
200222 2025-08-16 19:38:07+00:00     364000                  Fukurokouji   
200223 2025-08-16 19:41:14+00:00     186227                 Door Shutter   
200224 2025-08-16 19:46:19+00:00     304080                       Locked   

       master_metadata_album_artist_name  \
0                       Frederick Delius   
1                          Gabriel Fauré   
2                             Tom Browne   
3                           Ohio Players   
4                            James Brown   
...                                  ...   
200220                         El Huervo   
200221                          Ebi Soda   
200222                         El Huervo   
200223                       Brokenchord   
200224                           Dimlite   

               master_metadata_album_album_name  \
0                   Romance for Cello and Piano   
1       Fauré: Requiem, Op. 48 / Pavane, Op. 50   
2                           Funkin' For Jamaica   
3                                         Honey   
4                                   The Payback   
...                                         ...   
200220                              World's End   
200221                                      Ugh   
200222                              World's End   
200223                     Endless Transmission   
200224                        This Is Embracing   

                           spotify_track_uri reason_start reason_end  shuffle  \
0       spotify:track:70nMUxP46mAcjW5iRpRiyN     clickrow    endplay    False   
1       spotify:track:2oqgnxjyNYItifLYl9MzaW     clickrow    endplay    False   
2       spotify:track:0xbkxTqSaW5blsYgRXpB5I     clickrow  trackdone    False   
3       spotify:track:1uhUJ5uTuHYdbitv3IsITR    trackdone  trackdone    False   
4       spotify:track:4u4Kb4W5mWtn9oz4A7SKK6    trackdone    endplay    False   
...                                      ...          ...        ...      ...   
200220  spotify:track:771NtAeF8EiFO4D4CQG2ij    trackdone    backbtn     True   
200221  spotify:track:5PETadLT856KzshN8ArpWl      backbtn     fwdbtn     True   
200222  spotify:track:771NtAeF8EiFO4D4CQG2ij       fwdbtn  trackdone     True   
200223  spotify:track:2bBrirDuZf2mDbE6P7PCwh    trackdone  trackdone     True   
200224  spotify:track:3To51zCuaPkUK9HoUOXHfI    trackdone  trackdone     True   

        skipped  
0         False  
1         False  
2         False  
3         False  
4         False  
...         ...  
200220     True  
200221     True  
200222    False  
200223    False  
200224    False  

[200225 rows x 10 columns]

## Creation of derived columns

Now that my data feels clean and put together, I want to add more columns that will make analysis and visualisation easier and possible.

## Time-Based Derived Columns

The first set of derived columns I will create will be the following time-based columns:

- date
- year
- month
- year_month
- hour
- day_of_week
- day_name

In [24]:
print(f"The datetime in concern is {audio_df.loc[0, "ts"]}.")
print("")

# date
audio_df["date"] = audio_df["ts"].dt.date
print("date")
print(audio_df.loc[0, "date"])
print(f"The type of a \"date\" value is {type(audio_df.loc[0, "date"])}")
print(f"The Dtype of a \"date\" column is {audio_df["date"].dtype}")
print("")

# year
audio_df["year"] = audio_df["ts"].dt.year
print("year")
print(audio_df.loc[0, "year"])
print(f"The type of a \"year\" value is {type(audio_df.loc[0, "year"])}")
print(f"The Dtype of a \"year\" column is {audio_df["year"].dtype}")
print("")

# month
audio_df["month"] = audio_df["ts"].dt.month
print("month")
print(audio_df.loc[0, "month"])
print(f"The type of a \"month\" value is {type(audio_df.loc[0, "month"])}")
print(f"The Dtype of a \"month\" column is {audio_df["month"].dtype}")
print("")

# year_month

audio_df["year_month"] = audio_df["ts"].dt.to_period("M")
print("year_month")
print(audio_df.loc[0, "year_month"])
print(f"The type of a \"year_month\" value is {type(audio_df.loc[0, "year_month"])}")
print(f"The Dtype of a \"year_month\" column is {audio_df["year_month"].dtype}")
print("")

# hour

audio_df["hour"] = audio_df["ts"].dt.hour
print("hour")
print(audio_df.loc[0, "hour"])
print(f"The type of a \"hour\" value is {type(audio_df.loc[0, "hour"])}")
print(f"The Dtype of a \"hour\" column is {audio_df["hour"].dtype}")
print("")

# day_of_week
audio_df["day_of_week"] = audio_df["ts"].dt.dayofweek
print("day_of_week")
print(audio_df.loc[0, "day_of_week"])
print(f"The type of a \"day_of_week\" value is {type(audio_df.loc[0, "day_of_week"])}")
print(f"The Dtype of a \"day_of_week\" column is {audio_df["day_of_week"].dtype}")
print("")

# day_name
audio_df["day_name"] = audio_df["ts"].dt.day_name()
print("day_name")
print(audio_df.loc[0, "day_name"])
print(f"The type of a \"day_name\" value is {type(audio_df.loc[0, "day_name"])}")
print(f"The Dtype of a \"day_name\" column is {audio_df["day_name"].dtype}")
print("")


The datetime in concern is 2017-04-01 17:43:10+00:00.

date
2017-04-01
The type of a "date" value is <class 'datetime.date'>
The Dtype of a "date" column is object

year
2017
The type of a "year" value is <class 'numpy.int32'>
The Dtype of a "year" column is int32

month
4
The type of a "month" value is <class 'numpy.int32'>
The Dtype of a "month" column is int32

year_month
2017-04
The type of a "year_month" value is <class 'pandas._libs.tslibs.period.Period'>
The Dtype of a "year_month" column is period[M]

hour
17
The type of a "hour" value is <class 'numpy.int32'>
The Dtype of a "hour" column is int32

day_of_week
5
The type of a "day_of_week" value is <class 'numpy.int32'>
The Dtype of a "day_of_week" column is int32

day_name
Saturday
The type of a "day_name" value is <class 'str'>
The Dtype of a "day_name" column is object



/tmp/ipython-input-2266756020.py:30: UserWarning: Converting to PeriodArray/Index representation will drop timezone information.
  audio_df["year_month"] = audio_df["ts"].dt.to_period("M")


## Playback & Behaviour Columns

The second set of derived columns I will create will be the following playback & behaviour columns:

- listening_hours
    - Listening duration converted from milliseconds to hours
- track_play_count
    - Represents one play per row, used to count plays
- skip_flag
    - Numeric version of skip (1 = skipped, 0 = not skipped)
- completed_flag
    - Completion indicator (1 = not skipped, 0 = skipped)
- valid_play
    - Whether a play exceeds a 30 second duration threshold

In [25]:
# listening_hours

audio_df["listening_hours"] = audio_df["ms_played"] / 3_600_000
print(f"ms_played of first record is {audio_df.loc[0, "ms_played"]}")
print(f"listening_hours of first record is {audio_df.loc[0, "listening_hours"]}")
print("")

# track_play_count

audio_df["track_play_count"] = 1
display(audio_df["track_play_count"].head(3))

# skip_flag

audio_df["skip_flag"] = audio_df["skipped"].astype(int)
display(audio_df[["skipped","skip_flag"]])

# completed_flag

audio_df["completed_flag"] = 1 - audio_df["skip_flag"]
display(audio_df[["skipped","completed_flag"]])

# valid_play

audio_df["valid_play"] = audio_df["ms_played"] >= 30_000
display(audio_df[["ms_played","valid_play"]])

ms_played of first record is 36860
listening_hours of first record is 0.01023888888888889



0    1
1    1
2    1
Name: track_play_count, dtype: int64

skipped  skip_flag
0         False          0
1         False          0
2         False          0
3         False          0
4         False          0
...         ...        ...
200220     True          1
200221     True          1
200222    False          0
200223    False          0
200224    False          0

[200225 rows x 2 columns]

skipped  completed_flag
0         False               1
1         False               1
2         False               1
3         False               1
4         False               1
...         ...             ...
200220     True               0
200221     True               0
200222    False               1
200223    False               1
200224    False               1

[200225 rows x 2 columns]

ms_played  valid_play
0           36860        True
1           36800        True
2          280933        True
3          290160        True
4          459533        True
...           ...         ...
200220      36896        True
200221       4563       False
200222     364000        True
200223     186227        True
200224     304080        True

[200225 rows x 2 columns]

## Displaying the derived columns

I will now display the derived columns:

In [26]:
derived_columns = [
    "date",
    "year",
    "month",
    "year_month",
    "hour",
    "day_of_week",
    "day_name",
    "listening_hours",
    "track_play_count",
    "skip_flag",
    "completed_flag",
    "valid_play"
]

audio_df[derived_columns]

date  year  month year_month  hour  day_of_week  day_name  \
0       2017-04-01  2017      4    2017-04    17            5  Saturday   
1       2017-04-01  2017      4    2017-04    19            5  Saturday   
2       2017-04-01  2017      4    2017-04    19            5  Saturday   
3       2017-04-01  2017      4    2017-04    20            5  Saturday   
4       2017-04-01  2017      4    2017-04    20            5  Saturday   
...            ...   ...    ...        ...   ...          ...       ...   
200220  2025-08-16  2025      8    2025-08    19            5  Saturday   
200221  2025-08-16  2025      8    2025-08    19            5  Saturday   
200222  2025-08-16  2025      8    2025-08    19            5  Saturday   
200223  2025-08-16  2025      8    2025-08    19            5  Saturday   
200224  2025-08-16  2025      8    2025-08    19            5  Saturday   

        listening_hours  track_play_count  skip_flag  completed_flag  \
0              0.010239                 1          0               1   
1              0.010222                 1          0               1   
2              0.078037                 1          0               1   
3              0.080600                 1          0               1   
4              0.127648                 1          0               1   
...                 ...               ...        ...             ...   
200220         0.010249                 1          1               0   
200221         0.001267                 1          1               0   
200222         0.101111                 1          0               1   
200223         0.051730                 1          0               1   
200224         0.084467                 1          0               1   

        valid_play  
0             True  
1             True  
2             True  
3             True  
4             True  
...            ...  
200220        True  
200221       False  
200222        True  
200223        True  
200224        True  

[200225 rows x 12 columns]

## Writing processed data to CSV file

I feel that the data is usable for analysis and visualisation now in Tableau. I understand that the data's quality can be assessed more, and the code can be tidied up, but due to time constraints, this will be work I do in the future. I will now write the data to a CSV file:

In [ ]:
processed_location = "data/processed/processed_spotify_data.csv"
audio_df.to_csv(processed_location, encoding="utf-8")